### PR TITLE
feat(typescript-estree): align optional fields

### LINKED ISSUE #1097: Missing `optional` when the key is `MemberExpression` in a computed optional methods.
This issue is related on https://github.com/prettier/prettier/pull/6601#issuecomment-541992951

**What code were you trying to parse?**

```ts
class Foo {
  [foo.bar]?() {}
}
```

**What did you expect to happen?**

```js
{
  "type": "Program",
  "body": [
    {
      "type": "ClassDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Foo",
        "range": [
          6,
          9
        ],
        "loc": {
          "start": {
            "line": 1,
            "column": 6
          },
          "end": {
            "line": 1,
            "column": 9
          }
        }
      },
      "body": {
        "type": "ClassBody",
        "body": [
          {
            "type": "MethodDefinition",
            "key": {
              "type": "MemberExpression",
              "object": {
                "type": "Identifier",
                "name": "foo",
                "range": [
                  17,
                  20
                ],
                "loc": {
                  "start": {
                    "line": 2,
                    "column": 5
                  },
                  "end": {
                    "line": 2,
                    "column": 8
                  }
                }
              },
              "property": {
                "type": "Identifier",
                "name": "bar",
                "range": [
                  21,
                  24
                ],
                "loc": {
                  "start": {
                    "line": 2,
                    "column": 9
                  },
                  "end": {
                    "line": 2,
                    "column": 12
                  }
                }
              },
              "computed": false,
              "optional": true, // <= this `optional` should be `true`
              "range": [
                17,
                24
              ],
              "loc": {
                "start": {
                  "line": 2,
                  "column": 5
                },
                "end": {
                  "line": 2,
                  "column": 12
                }
              }
            },
            "value": {
              "type": "FunctionExpression",
              "id": null,
              "generator": false,
              "expression": false,
              "async": false,
              "body": {
                "type": "BlockStatement",
                "body": [],
                "range": [
                  29,
                  31
                ],
                "loc": {
                  "start": {
                    "line": 2,
                    "column": 17
                  },
                  "end": {
                    "line": 2,
                    "column": 19
                  }
                }
              },
              "range": [
                26,
                31
              ],
              "params": [],
              "loc": {
                "start": {
                  "line": 2,
                  "column": 14
                },
                "end": {
                  "line": 2,
                  "column": 19
                }
              }
            },
            "computed": true,
            "static": false,
            "kind": "method",
            "range": [
              16,
              31
            ],
            "loc": {
              "start": {
                "line": 2,
                "column": 4
              },
              "end": {
                "line": 2,
                "column": 19
              }
            }
          }
        ],
        "range": [
          10,
          33
        ],
        "loc": {
          "start": {
            "line": 1,
            "column": 10
          },
          "end": {
            "line": 3,
            "column": 1
          }
        }
      },
      "superClass": null,
      "range": [
        0,
        33
      ],
      "loc": {
        "start": {
          "line": 1,
          "column": 0
        },
        "end": {
          "line": 3,
          "column": 1
        }
      }
    }
  ],
  "sourceType": "script",
  "range": [
    0,
    34
  ],
  "loc": {
    "start": {
      "line": 1,
      "column": 0
    },
    "end": {
      "line": 4,
      "column": 0
    }
  }
}
```

**What actually happened?**

```js
{
  "type": "Program",
  "body": [
    {
      "type": "ClassDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Foo",
        "range": [
          6,
          9
        ],
        "loc": {
          "start": {
            "line": 1,
            "column": 6
          },
          "end": {
            "line": 1,
            "column": 9
          }
        }
      },
      "body": {
        "type": "ClassBody",
        "body": [
          {
            "type": "MethodDefinition",
            "key": {
              "type": "MemberExpression",
              "object": {
                "type": "Identifier",
                "name": "foo",
                "range": [
                  17,
                  20
                ],
                "loc": {
                  "start": {
                    "line": 2,
                    "column": 5
                  },
                  "end": {
                    "line": 2,
                    "column": 8
                  }
                }
              },
              "property": {
                "type": "Identifier",
                "name": "bar",
                "range": [
                  21,
                  24
                ],
                "loc": {
                  "start": {
                    "line": 2,
                    "column": 9
                  },
                  "end": {
                    "line": 2,
                    "column": 12
                  }
                }
              },
              "computed": false,
              "optional": false, // <= this `optional` should be `true`
              "range": [
                17,
                24
              ],
              "loc": {
                "start": {
                  "line": 2,
                  "column": 5
                },
                "end": {
                  "line": 2,
                  "column": 12
                }
              }
            },
            "value": {
              "type": "FunctionExpression",
              "id": null,
              "generator": false,
              "expression": false,
              "async": false,
              "body": {
                "type": "BlockStatement",
                "body": [],
                "range": [
                  29,
                  31
                ],
                "loc": {
                  "start": {
                    "line": 2,
                    "column": 17
                  },
                  "end": {
                    "line": 2,
                    "column": 19
                  }
                }
              },
              "range": [
                26,
                31
              ],
              "params": [],
              "loc": {
                "start": {
                  "line": 2,
                  "column": 14
                },
                "end": {
                  "line": 2,
                  "column": 19
                }
              }
            },
            "computed": true,
            "static": false,
            "kind": "method",
            "range": [
              16,
              31
            ],
            "loc": {
              "start": {
                "line": 2,
                "column": 4
              },
              "end": {
                "line": 2,
                "column": 19
              }
            }
          }
        ],
        "range": [
          10,
          33
        ],
        "loc": {
          "start": {
            "line": 1,
            "column": 10
          },
          "end": {
            "line": 3,
            "column": 1
          }
        }
      },
      "superClass": null,
      "range": [
        0,
        33
      ],
      "loc": {
        "start": {
          "line": 1,
          "column": 0
        },
        "end": {
          "line": 3,
          "column": 1
        }
      }
    }
  ],
  "sourceType": "script",
  "range": [
    0,
    34
  ],
  "loc": {
    "start": {
      "line": 1,
      "column": 0
    },
    "end": {
      "line": 4,
      "column": 0
    }
  }
}
```

**Versions**

| package                                | version   |
| -------------------------------------- | --------- |
| `@typescript-eslint/typescript-estree` | `2.4.0`   |
| `TypeScript`                           | `3.6.4`   |
| `node`                                 | `10.16.3` |
| `npm`                                  | `6.9.0`   |

**Note**

This bug does not happen with simple identifer.

```ts
class Foo {
  [foo]?() {}
}
```

### PR DESCRIPTION
**Pull request by armano2**
**This is BREAKING CHANGE**

Goal of this change is to align how we are marking optional properties and methods, 
Difference between this PR and #1100 is that instead of trying to add "optional" to key of method/property we are directly marking method with it.

This is more aligned with how typescript produces ast and how babel generates estree for this field.

read more: https://github.com/typescript-eslint/typescript-eslint/pull/1100#discussion_r336752956
context https://github.com/prettier/prettier/pull/6601, https://github.com/prettier/prettier/issues/6678

closes #1100
fixes #1097

cc @sosukesuzuki  @thorn0 @fisker

### CODE REVIEW

File: packages/typescript-estree/src/ts-estree/ts-estree.ts [hunk 1]
```diff
@@ -641,26 +643,26 @@ interface TSMethodSignatureBase extends BaseNode {
...
   typeAnnotation?: TSTypeAnnotation;
   initializer?: Expression;
+  accessibility?: Accessibility;
   readonly?: boolean;
-  static?: boolean;
   export?: boolean;
-  accessibility?: Accessibility;
+  static?: boolean;
+  optional?: boolean;
 }
```
Comment (bradzacher) [inline packages/typescript-estree/src/ts-estree/ts-estree.ts]: nitpick - just to reduce diff noise, could you undo this sorting?

**Comment by bradzacher** [comment]
why is this breaking? It looks like a backwards-compatible addition.

**Comment by armano2** [comment]
AST produced by our conversion is going to differ for existing code base

**Comment by bradzacher** [comment]
As long as a consumer's runtime code doesn't break as part of the addition, it's non-breaking.
In this instance:
- the fields are optional
- no fields are being removed

So the overall shape of the AST is consistent, and existing code will not break.
Breaking snapshot tests is not included in consideration for breaking changes for obvious reasons (or else every new TS feature would require a major bump).

**Comment by thorn0** [comment]
> why is this breaking? It looks like a backwards-compatible addition.

Now, in a `MethodDefinition` node, the key is marked with `optional: true` (only if the key is an identifier). After this change it won't be marked.

**Comment by bradzacher** [comment]
Ah, sorry - I missed the missing `.key` in `result.optional`.

Question:
Did we want to do both in the interim?
I imagine it'll be a bit clunky if prettier has to support both versions.
Could ease the transition if we kept the existing functionality in for a major version.

**Comment by thorn0** [comment]
Don't worry about Prettier. It's already smart enough to find `optional` on the method definition node.

**Comment by armano2** [comment]
@bradzacher i will prefer to not replicate / duplicate informations across nodes, thats why its a breaking change. 
prettier team (namely @thorn0) started implementing alternative ts parser (babel) and they have to support this change anyway as babel produces ast in this way.

**Pull request merged**